### DISCUSSION TITLE
Bunch of improvement requests/ suggestions

### DISCUSSION THREAD

**Discussion by Arkhorse** [bcssov/IronyModManager#213] [Ideas]
I'm going to forgo the template, as it doesn't really fit here.

First, I am a contributor to a project called "Relhax Modpack". This modpack is in CS code, so its code may help you guys out. The Dev/s are professionals, so maybe something good there. Link: https://github.com/Willster419/RelhaxModpack. It is open source, but if you have any questions, join our discord and ask Willster, or just post an issue. Note: I don't personally do much in the way of programing the application, I handle the database side of it along with others.

Now on to the actual requests/ suggestions

- The Export and Import buttons could be changed to all have their own buttons. There is more than enough room up there to do this. And with dynamic resizing of buttons, would support the app being in different resolutions
- Add a "Refresh" Button to refresh the apps mod definitions, so the app can import newly downloaded or created mods without needing to relaunch the app. EDIT: Ignore this, I didnt see the button in the right click menu. Maybe move it to beside the actions button?
    - Simply importing things tends to break the patch mods. Guess this is a bug.
- Add the ability to contribute compatibility patches to a database to others can use them instead of making their own. These could be saved in xml format, and pulled/pushed to the GitHub or a ftp server (wouldn't need to be a large one)
    - If you use GitHub, you could use discord for users to upload their own patches, or simply a google form.
- More obvious conflict wins. Sometimes its hard to tell exactly which mod is winning the conflict. Maybe a simple check mark would suffice.
- Save ignore rules. This one is rather annoying. I have to constantly remake any ignore rules I put in either after an app relaunch, or with other profiles/ collections.
- Load Order Optimization Tool (Like LOOT from the Fallout games). Like with the mod compatibility patch database, there could be a system to auto sort mods to where they need to be, and leave any mod not known at the top for the user to manually sort.
    - Loot also has the ability to know if a user should have certain patch mods installed, and link them to the user. Could also be useful
 - The ability to do something like the ignore rule modName, but altered. This rule would simply say that this mod should always lose, or this mod should always win.
     - District Overhaul 2 would be an always win, !!Ariphaos Unofficial Patch would always lose.

Well, thats all I can think of right now. Will update this with another section if I think of anymore.

**Comment by bcssov** [comment]
> This modpack is in CS code, so its code may help you guys out. The Dev/s are professionals, so maybe something good there

So am I :)  
But honestly it's difficult to use someone else’s code designed for a completely different reason in another unrelated app. But thanks anyway.

> The Export and Import buttons could be changed to all have their own buttons. There is more than enough room up there to do this. And with dynamic resizing of buttons, would support the app being in different resolutions

While I'm no UX expert myself obviously (I'm a developer). I'd prefer to keep the buttons as they are, I could develop various viewports to keep a "responsive" philosophy in place here, so the layout would change. But given how now Irony supports multiple languages and being very easy to add another language, that'll mean too many problems for me down the line. Even now there are problems as text in some of the localizations is too wide. Secondly if the buttons are stacked in more rows this will mean less space for mod management as the listbox where the mods are located would lose some space.

> Simply importing things tends to break the patch mods. Guess this is a bug.

Too little information for me to reply. I was not made aware of any existing issues in that department. What I could possibly see potentially confusing you is Irony stating that conflict solver needs a rerun each time the collection (playset) is changed. That is intentional and not an issue or a bug. While Irony does not do any auto patching it does do some automated fixing behind the scenes and any change in the load order potentially requires a conflict solver re-run.

> Add the ability to contribute compatibility patches to a database to others can use them instead of making their own. These could be saved in xml format, and pulled/pushed to the GitHub or a ftp server (wouldn't need to be a large one) 

That was actually a plan but given how I'm a one man army I had to give up on that idea. My main inspiration to how this would work would have been CKAN\NETKAN for Kerbal Space Program.

> More obvious conflict wins. Sometimes its hard to tell exactly which mod is winning the conflict. Maybe a simple check mark would suffice.

Irony already tells you what will win and why.
- Orange borders around an item indicate what will be used by default by the game
- Green borders indicate a virtual definition which, if resolved (diff merged) will override anything that's listed. The virtual definition uses orange definition content as the starting point. The idea behind it is to do diff merging with all the definitions into the green one.

> Save ignore rules. This one is rather annoying. I have to constantly remake any ignore rules I put in either after an app relaunch, or with other profiles/ collections.

I opted against the global ignore rules logic as each mod collection is different, so they were made per mod patch. If you delete patches the ignore rules will be lost, same as if you have multiple mod collections which have their own mod patches they will not be synced as they are per mod patch (per collection).

> Load Order Optimization Tool (Like LOOT from the Fallout games). Like with the mod compatibility patch database, there could be a system to auto sort mods to where they need to be, and leave any mod not known at the top for the user to manually sort.

It's also has been a point of discussion on the Irony's stellaris modding den channel. The problem is that very little to no people are willing to help on this one. Thus such a feature would be a waste of time and resources I can dedicate to this project (of which I have very little time as it is at the moment). Simply putting it it's hard to motivate people of Stellaris community for some reason. I also planned to borrow ideas from how patching is done from KSP ModuleManager onto stellaris (it would of course happen before launching the game) but that too was cut for the same reason.

> The ability to do something like the ignore rule modName, but altered. This rule would simply say that this mod should always lose, or this mod should always win. 

I like that. 

After more info is gathered this will have to be broken down into smaller sub sections, I'm not a fan of a single giant issue with various unrelated topics.

EDIT: Converted this into a discussion as it's more appropriate.

**Reply by Arkhorse** [reply]
> > This modpack is in CS code, so its code may help you guys out. The Dev/s are professionals, so maybe something good there
> 
> So am I :)
> But honestly it's difficult to use someone else’s code designed for a completely different reason in another unrelated app. But thanks anyway.

The modpack is very similar to this. Manages mods in all ways (or will once the coding is done). But I understand what you are saying. I did some modding in wot with python, and alot of open source mods couldnt be used either due to a different style (making it hard to read), or was poorly coded (not talking about that here, just an example).
 
> > The Export and Import buttons could be changed to all have their own buttons. There is more than enough room up there to do this. And with dynamic resizing of buttons, would support the app being in different resolutions
> 
> While I'm no UX expert myself obviously (I'm a developer). I'd prefer to keep the buttons as they are, I could develop various viewports to keep a "responsive" philosophy in place here, so the layout would change. But given how now Irony supports multiple languages and being very easy to add another language, that'll mean too many problems for me down the line. Even now there are problems as text in some of the localizations is too wide. Secondly if the buttons are stacked in more rows this will mean less space for mod management as the listbox where the mods are located would lose some space.

Yeah, I guess changing how those buttons work would require a bit of a UX/UI change to accommodate the other languages. Could be a purpose built button row/s and the combobox (listbox...) for the mod management could be moved to have its own line with the two buttons (+/-) on either side, or as it is now. But, I am not sure how that would affect adding more languages down the line. Do you use wpf for the ui? If so, I can have a look at doing this for you. Thats the bit I do for relhax as well as database management

> 
> > Simply importing things tends to break the patch mods. Guess this is a bug.
> 
> Too little information for me to reply. I was not made aware of any existing issues in that department. What I could possibly see potentially confusing you is Irony stating that conflict solver needs a rerun each time the collection (playset) is changed. That is intentional and not an issue or a bug. While Irony does not do any auto patching it does do some automated fixing behind the scenes and any change in the load order potentially requires a conflict solver re-run.

Uh... What happens for me, is I would hit import from paradox launcher to import new mods (now I know there is a dedicated button for that), and the mod collections patch mod would be lost. The files still exist, but for some reason, IMM cant find it. 

> 
> > Add the ability to contribute compatibility patches to a database to others can use them instead of making their own. These could be saved in xml format, and pulled/pushed to the GitHub or a ftp server (wouldn't need to be a large one)
> 
> That was actually a plan but given how I'm a one man army I had to give up on that idea. My main inspiration to how this would work would have been CKAN\NETKAN for Kerbal Space Program.

Yeah, CKAN is great. But being a one man army for something like this is hard. Have you posted anything on the forums or on steam asking for help? Never know...

> 
> > More obvious conflict wins. Sometimes its hard to tell exactly which mod is winning the conflict. Maybe a simple check mark would suffice.
> 
> Irony already tells you what will win and why.
> 
> * Orange borders around an item indicate what will be used by default by the game
> * Green borders indicate a virtual definition which, if resolved (diff merged) will override anything that's listed. The virtual definition uses orange definition content as the starting point. The idea behind it is to do diff merging with all the definitions into the green one.

This still doesnt help me. When I am going through the conflicts, there are two colours: The orange box one, and a blue filled box. Ive never seen the green borders one.

> 
> > Save ignore rules. This one is rather annoying. I have to constantly remake any ignore rules I put in either after an app relaunch, or with other profiles/ collections.
> 
> I opted against the global ignore rules logic as each mod collection is different, so they were made per mod patch. If you delete patches the ignore rules will be lost, same as if you have multiple mod collections which have their own mod patches they will not be synced as they are per mod patch (per collection).

Yeah, different mod collections will have different needs. But this function could act like a base. Some mods I always want to ignore, same with some folders. Right now I have that saved to a txt file, and I have to copy paste it into the ignore rules everytime.

> 
> > Load Order Optimization Tool (Like LOOT from the Fallout games). Like with the mod compatibility patch database, there could be a system to auto sort mods to where they need to be, and leave any mod not known at the top for the user to manually sort.
> 
> It's also has been a point of discussion on the Irony's stellaris modding den channel. The problem is that very little to no people are willing to help on this one. Thus such a feature would be a waste of time and resources I can dedicate to this project (of which I have very little time as it is at the moment). Simply putting it it's hard to motivate people of Stellaris community for some reason. I also planned to borrow ideas from how patching is done from KSP ModuleManager onto stellaris (it would of course happen before launching the game) but that too was cut for the same reason.
> 
> > The ability to do something like the ignore rule modName, but altered. This rule would simply say that this mod should always lose, or this mod should always win.
> 
> I like that.
> 
> After more info is gathered this will have to be broken down into smaller sub sections, I'm not a fan of a single giant issue with various unrelated topics.
> 
> EDIT: Converted this into a discussion as it's more appropriate.

**Reply by bcssov** [reply]
> The modpack is very similar to this. Manages mods in all ways (or will once the coding is done). But I understand what you are saying. I did some modding in wot with python, and alot of open source mods couldnt be used either due to a different style (making it hard to read), or was poorly coded (not talking about that here, just an example).

Irony is not a modpack but a mod manager, which are 2 different things in this case (I think). I took ideas and design philosophies from other Stellaris related projects like SCModManager, Paradoxos and even Paradox's own Launcher. I'm all for reusing code if it'll help me speed up some process. I don't reinvent the wheel if I can find a suitable nuget library or an open source code (with a suitable license - MIT, Apache...) to use in this project.

> Yeah, I guess changing how those buttons work would require a bit of a UX/UI change to accommodate the other languages. Could be a purpose built button row/s and the combobox (listbox...) for the mod management could be moved to have its own line with the two buttons (+/-) on either side, or as it is now. But, I am not sure how that would affect adding more languages down the line. 

That's why I need a UX expert to do a redesign, just mish mashing around will not really work. So I'd prefer to leave it like that until someone with UX experience can provide some proposed mockups for me. If you really wanna take a crack at it download Pencil or use some other wireframing software of your choice and shoot me up some proposals.

> Do you use wpf for the ui? If so, I can have a look at doing this for you. Thats the bit I do for relhax as well as database management

Since this project is cross platform I am not using WPF. At the time I started working on Irony I had a few framework choices to create a cross platform solution (net maui was not really a thing a year ago). Out of all of these I opted to use Avalonia which a cross platform XAML framework, but it has its own controls and its own set of problems. Any existing libraries that I might want to use I have to port to Avalonia. And some I had to make on my own (like drag and drop of mods). The only thing in common between WPF and Avalonia is that they both use XAML as the markup language.

> Uh... What happens for me, is I would hit import from paradox launcher to import new mods (now I know there is a dedicated button for that), and the mod collections patch mod would be lost. The files still exist, but for some reason, IMM cant find it.

Importing from paradox launcher is importing the load order of the selected playset (not new mods). And your patch is not lost but import creates a new mod collection under a new name (the matching playset name). This newly created mod collection is considered clean and does not have any patches tied to it. No bugs here that I can see.

> Yeah, CKAN is great. But being a one man army for something like this is hard. Have you posted anything on the forums or on steam asking for help? Never know...

Had few people volunteer only to give up, after a couple of these instances you kinda give up on looking. Like I said it's difficult to motivate Stellaris community.

> This still doesnt help me. When I am going through the conflicts, there are two colours: The orange box one, and a blue filled box. Ive never seen the green borders one.

Honestly, you're the first person who stated that. The green one only shows for text conflicts.

![1](https://user-images.githubusercontent.com/5864390/112206603-682ed180-8c16-11eb-8306-834e880bae8f.png)

> Yeah, different mod collections will have different needs. But this function could act like a base. Some mods I always want to ignore, same with some folders. Right now I have that saved to a txt file, and I have to copy paste it into the ignore rules everytime.

Average user only uses 1 or 2 mod collections to be honest.

**Reply by Arkhorse** [reply]
> > The modpack is very similar to this. Manages mods in all ways (or will once the coding is done). But I understand what you are saying. I did some modding in wot with python, and alot of open source mods couldnt be used either due to a different style (making it hard to read), or was poorly coded (not talking about that here, just an example).
> 
> Irony is not a modpack but a mod manager, which are 2 different things in this case (I think). I took ideas and design philosophies from other Stellaris related projects like SCModManager, Paradoxos and even Paradox's own Launcher. I'm all for reusing code if it'll help me speed up some process. I don't reinvent the wheel if I can find a suitable nuget library or an open source code (with a suitable license - MIT, Apache...) to use in this project.

Right.... In wot, modpacks and mod managers are considered the same thing. Relhax is a mod manager. The most advanced one for wot (99% of all the others use inno setup 🤢). It has a patching engine, database system, and alot of other things that might be useful. 

> > Yeah, I guess changing how those buttons work would require a bit of a UX/UI change to accommodate the other languages. Could be a purpose built button row/s and the combobox (listbox...) for the mod management could be moved to have its own line with the two buttons (+/-) on either side, or as it is now. But, I am not sure how that would affect adding more languages down the line.
> 
> That's why I need a UX expert to do a redesign, just mish mashing around will not really work. So I'd prefer to leave it like that until someone with UX experience can provide some proposed mockups for me. If you really wanna take a crack at it download Pencil or use some other wireframing software of your choice and shoot me up some proposals.
> 

Ok, ill have to read into those programs

> > Do you use wpf for the ui? If so, I can have a look at doing this for you. Thats the bit I do for relhax as well as database management
> 
> Since this project is cross platform I am not using WPF. At the time I started working on Irony I had a few framework choices to create a cross platform solution (net maui was not really a thing a year ago). Out of all of these I opted to use Avalonia which a cross platform XAML framework, but it has its own controls and its own set of problems. Any existing libraries that I might want to use I have to port to Avalonia. And some I had to make on my own (like drag and drop of mods). The only thing in common between WPF and Avalonia is that they both use XAML as the markup language.
> 
> > Uh... What happens for me, is I would hit import from paradox launcher to import new mods (now I know there is a dedicated button for that), and the mod collections patch mod would be lost. The files still exist, but for some reason, IMM cant find it.
> 
> Importing from paradox launcher is importing the load order of the selected playset (not new mods). And your patch is not lost but import creates a new mod collection under a new name (the matching playset name). This newly created mod collection is considered clean and does not have any patches tied to it. No bugs here that I can see.
> 

Ah ok.

> > Yeah, CKAN is great. But being a one man army for something like this is hard. Have you posted anything on the forums or on steam asking for help? Never know...
> 
> Had few people volunteer only to give up, after a couple of these instances you kinda give up on looking. Like I said it's difficult to motivate Stellaris community.
> 

Yeah, same in wot

> > This still doesnt help me. When I am going through the conflicts, there are two colours: The orange box one, and a blue filled box. Ive never seen the green borders one.
> 
> Honestly, you're the first person who stated that. The green one only shows for text conflicts.
> 
> ![1](https://user-images.githubusercontent.com/5864390/112206603-682ed180-8c16-11eb-8306-834e880bae8f.png)
> 

So the blue box doesnt do anything?

> > Yeah, different mod collections will have different needs. But this function could act like a base. Some mods I always want to ignore, same with some folders. Right now I have that saved to a txt file, and I have to copy paste it into the ignore rules everytime.
> 
> Average user only uses 1 or 2 mod collections to be honest.

Oh... I make a new one for every empire I create....

**Reply by bcssov** [reply]
I've never played WOT so I'm rather ignorant of the community and stuff that happens there.

> Ok, ill have to read into those programs

For initial wireframing I used Pencil, if you don't want to start from scratch you can use this as a template: https://github.com/bcssov/IronyModManager/blob/master/Wireframes/IronyModManager.epgz
Just beware that the wireframe is dangerously outdated.

> So the blue box doesnt do anything?

Blue is an indicator of what code is shown in the diff merge control below.

![1](https://user-images.githubusercontent.com/5864390/112211065-c01c0700-8c1b-11eb-8782-59ccfffd5161.png)

When you hit resolve the content of the green box is dumpled into the patch mod. I think that is somewhat covered in the wiki (again I'm not much of a document writer as well). What you're looking at here is a quasi n level merge situation. Depending on the number of duplicates there are detected.

**Reply by Arkhorse** [reply]
> I've never played WOT so I'm rather ignorant of the community and stuff that happens there.

Yeah, its an ... odd game. And an even odder community. Almost all mods are made by the Russian region of the game... But yeah, Relhax's code is under the `Apache License`, so you should be able to use any of it. 

> 
> > So the blue box doesnt do anything?
> 
> Blue is an indicator of what code is shown in the diff merge control below.
> 
> ![1](https://user-images.githubusercontent.com/5864390/112211065-c01c0700-8c1b-11eb-8782-59ccfffd5161.png)
> 
> When you hit resolve the content of the green box is dumpled into the patch mod. I think that is somewhat covered in the wiki (again I'm not much of a document writer as well). What you're looking at here is a quasi n level merge situation. Depending on the number of duplicates there are detected.

Ok, but it doesnt tell me which mod the green content is from... Unless it does, and I am just not seeing it :P

**Reply by bcssov** [reply]
> Yeah, its an ... odd game. And an even odder community. Almost all mods are made by the Russian region of the game... But yeah, Relhax's code is under the Apache License, so you should be able to use any of it.

If I use any code it's for the UI. But since the UI frameworks are different, there comes nothing to mind I could use.

> Ok, but it doesnt tell me which mod the green content is from... Unless it does, and I am just not seeing it :P

Documentation covers this area:
1. If there's nothing in resolution history then the content initially is set from the orange colored box
2. If you previously resolved then reset the conflict. Or the conflict needs attention once again due to the changes to the code for example. The archived code is used from the last time you resolved it aka conflict history. If you dive into the patch mod folder and look into state_conflict_history directory you'll see that that's where the resolution history is kept.

**Reply by Arkhorse** [reply]
I watched that video pinned on the discord. Explains it in a way I can understand. Dont know why I was stumbling over it on the wiki :P

**Reply by bcssov** [reply]
I'm open to improvements to the documentation on the wiki, like I said I'm not a documentation writer. I'll add you as a contributor if you want it. I keep the wiki locked so I don't have to moderate it in case anyone asks.

**Reply by Arkhorse** [reply]
> I'm open to improvements to the documentation on the wiki, like I said I'm not a documentation writer. I'll add you as a contributor if you want it. I keep the wiki locked so I don't have to moderate it in case anyone asks.

I can attempt to help you with it. But I just want to note that my time is ... erratic. Personal issues :P. I do write most of the documentation for that modpack I posted, other than the application documentation, thats done by willster...

**Reply by bcssov** [reply]
Well you got a collaborator invite now.
